sql: TestTrace failed

sql.TestTrace [failed](https://mesolite.cluster.engflow.com/invocations/default/913f1733-e94d-49de-b251-c8fe2a31df44?testReportRun=1&testReportShard=14&testReportAttempt=1#targets-Ly9wa2cvc3FsOnNxbF90ZXN0) on master @ [dba914a42e73cbb22ca24dcc58fc89fde8be87a6](https://github.com/cockroachdb/cockroach/commits/dba914a42e73cbb22ca24dcc58fc89fde8be87a6):


```
        read-only path
        sequencing request
        acquiring latches
        scanning lock table for conflicting locks
        can't serve follower read; closed timestamp too low by: 3.667483889s; maxClosed: 1764062586.914060984,0 ts: 1764062590.081544873,0 uncertaintyLimit: 1764062590.581544873,0
        replica.Send got error: [NotLeaseHolderError] lease held by different store; r79: replica (n3,s3):2 not lease holder; current lease is repl=(n1,s1):1 seq=1 start=0,0 type=LeaseEpoch epo=1 min-exp=0,0 pro=1764062577.298034422,0 acq=Request
        error from stores.Send: [NotLeaseHolderError] lease held by different store; r79: replica (n3,s3):2 not lease holder; current lease is repl=(n1,s1):1 seq=1 start=0,0 type=LeaseEpoch epo=1 min-exp=0,0 pro=1764062577.298034422,0 acq=Request: "sql txn" meta={id=6b6eed15 key=/Min iso=Serializable pri=0.02571349 epo=0 ts=1764062590.081544873,0 min=1764062590.081544873,0 seq=0} lock=false stat=PENDING rts=1764062590.081544873,0 gul=1764062590.581544873,0 obs={n3@1764062590.081544873,0}
        node sending response
        received batch response
        [NotLeaseHolderError] lease held by different store; r79: replica (n3,s3):2 not lease holder; current lease is repl=(n1,s1):1 seq=1 start=0,0 type=LeaseEpoch epo=1 min-exp=0,0 pro=1764062577.298034422,0 acq=Request: "sql txn" meta={id=6b6eed15 key=/Min iso=Serializable pri=0.02571349 epo=0 ts=1764062590.081544873,0 min=1764062590.081544873,0 seq=0} lock=false stat=PENDING rts=1764062590.081544873,0 gul=1764062590.581544873,0 obs={n3@1764062590.081544873,0}
        backing off due to NotLeaseHolderErr with stale info (updatedLH=false intentionallySentToFollower=false leaseholderUnavailable=false)
        will retry after 48.969039ms
        application error: [NotLeaseHolderError] lease held by different store; r79: replica (n3,s3):2 not lease holder; current lease is repl=(n1,s1):1 seq=1 start=0,0 type=LeaseEpoch epo=1 min-exp=0,0 pro=1764062577.298034422,0 acq=Request: "sql txn" meta={id=6b6eed15 key=/Min iso=Serializable pri=0.02571349 epo=0 ts=1764062590.081544873,0 min=1764062590.081544873,0 seq=0} lock=false stat=PENDING rts=1764062590.081544873,0 gul=1764062590.581544873,0 obs={n3@1764062590.081544873,0}
        trying next peer (n1,s1):1
        sending request to 127.0.0.1:40349
        sending batch request
        === SPAN START: /cockroach.roachpb.KVBatch/Batch ===
        === SPAN START: /cockroach.roachpb.Internal/Batch ===
        node received request: 1 Scan
        executing Scan [/Table/106/1,/Table/106/2), [txn: 6b6eed15], [can-forward-ts], [max_span_request_keys: 0], [target_bytes: 39009]
        read-only path
        sequencing request
        acquiring latches
        scanning lock table for conflicting locks
        can drop latches early for batch (Scan [/Table/106/1,/Table/106/2), [txn: 6b6eed15], [can-forward-ts], [max_span_request_keys: 0], [target_bytes: 39009]); scanning lock table first to detect conflicts
        lock table scan stats: seeked 1 times (1 internal); stepped 0 times (0 internal)
        lock table scan complete without conflicts; dropping latches early
        executing read-only batch
        evaluated Scan command header:<key:"\362\211" end_key:"\362\212" > scan_format:BATCH_RESPONSE , txn="sql txn" meta={id=6b6eed15 key=/Min iso=Serializable pri=0.02571349 epo=0 ts=1764062590.081544873,0 min=1764062590.081544873,0 seq=0} lock=false stat=PENDING rts=1764062590.081544873,0 gul=1764062590.581544873,0 obs={n1@1764062590.142210081,0 n3@1764062590.081544873,0} : resp=header:<> col_batches:<> , err=<nil>
        n1 scan stats: stepped 0 times (0 internal); seeked 1 times (1 internal); block-bytes: (total 0 B, cached 0 B, duration 0s); points: (count 0, key-bytes 0 B, value-bytes 0 B, tombstoned: 0) ranges: (count 0), (contained-points 0, skipped-points 0) evaluated requests: 0 gets, 1 scans, 0 reverse scans
        read completed
        node sending response
        received batch response
        recording span to refresh: /Table/106/{1-2}
        ComponentStats{ID: {39a16bf5-3a98-42ce-bc8a-b979ab432b45 PROCESSOR 0 3 }, KV nodes: n1, KV regions: test, KV time: 57ms, KV rows decoded: 0, KV bytes read: 0 B, KV gRPC calls: 1, KV pairs read: 0, MVCC step count (ext/int): 0/0, MVCC seek count (ext/int): 1/1, rows output: 0}
        ComponentStats{ID: {39a16bf5-3a98-42ce-bc8a-b979ab432b45 FLOW 0 3 test}}
        execution ends
        rows affected: 0
        === SPAN START: commit sql txn ===
        AutoCommit. err: <nil>
        releasing 4 descriptors
        release: 100("defaultdb",01018016169fe7564244ed9de88124afac465d) ver=1:0,0, refcount=0
        release: 104("test",01018016169fe7564244ed9de88124afac465d) ver=1:0,0, refcount=0
        release: 105("public",01018016169fe7564244ed9de88124afac465d) ver=1:0,0, refcount=0
        release: 106("foo",01018016169fe7564244ed9de88124afac465d) ver=1:0,0, refcount=0
        [NoTxn pos:10] executing ExecStmt: SET TRACING = off
        executing: SET TRACING = off in state: NoTxn
--- FAIL: TestTrace/ShowTraceFor/TracingOn/node-2 (0.13s)
=== RUN   TestTrace/ShowTraceForVectorized/TracingOff
--- FAIL: TestTrace/ShowTraceForVectorized/TracingOff (0.27s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=1</code>
 - <code>shard=14</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTrace.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-57241

Epic CRDB-55382

sql.TestTrace [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20777465?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20777465?buildTab=artifacts#/) on master @ [b2158da8f70c76ccf6570689824c505270c9cd2b](https://github.com/cockroachdb/cockroach/commits/b2158da8f70c76ccf6570689824c505270c9cd2b):


```
        executing Scan [/Tenant/10/Table/106/1,/Tenant/10/Table/106/2), [txn: 5406ea35], [can-forward-ts], [max_span_request_keys: 0], [target_bytes: 15188]
        read-only path
        sequencing request
        acquiring latches
        scanning lock table for conflicting locks
        can't serve follower read; closed timestamp too low by: 3.53784705s; maxClosed: 1764090352.232207212,0 ts: 1764090355.270054262,0 uncertaintyLimit: 1764090355.770054262,0
        replica.Send got error: [NotLeaseHolderError] lease held by different store; r92: replica (n2,s2):3 not lease holder; current lease is repl=(n1,s1):1 seq=1 start=0,0 type=LeaseEpoch epo=1 min-exp=0,0 pro=1764090346.740747622,0 acq=Request
        proxy request for Scan [/Tenant/10/Table/106/1,/Tenant/10/Table/106/2), [txn: 5406ea35], [can-forward-ts], [max_span_request_keys: 0], [target_bytes: 15188] after local error [NotLeaseHolderError] lease held by different store; r92: replica (n2,s2):3 not lease holder; current lease is repl=(n1,s1):1 seq=1 start=0,0 type=LeaseEpoch epo=1 min-exp=0,0 pro=1764090346.740747622,0 acq=Request: "sql txn" meta={id=5406ea35 key=/Min iso=Serializable pri=0.02132594 epo=0 ts=1764090355.270054262,0 min=1764090355.270054262,0 seq=0} lock=false stat=PENDING rts=1764090355.270054262,0 gul=1764090355.770054262,0 obs={n2@1764090355.271464799,0}
        === SPAN START: dist sender send ===
        processing a proxy request to repl=(n1,s1):1 seq=1 start=0,0 type=LeaseEpoch epo=1 min-exp=0,0 pro=1764090346.740747622,0 acq=Request
        clearing entries overlapping r92:/Tenant/10/Table/{48-107/1/5} [(n1,s1):1, (n3,s3):2, (n2,s2):3, next=4, gen=17]
        sending request to 127.0.0.1:32811
        sending batch request
        === SPAN START: /cockroach.roachpb.KVBatch/Batch ===
        === SPAN START: /cockroach.roachpb.Internal/Batch ===
        node received request: 1 Scan
        executing Scan [/Tenant/10/Table/106/1,/Tenant/10/Table/106/2), [txn: 5406ea35], [can-forward-ts], [max_span_request_keys: 0], [target_bytes: 15188]
        read-only path
        sequencing request
        acquiring latches
        scanning lock table for conflicting locks
        can drop latches early for batch (Scan [/Tenant/10/Table/106/1,/Tenant/10/Table/106/2), [txn: 5406ea35], [can-forward-ts], [max_span_request_keys: 0], [target_bytes: 15188]); scanning lock table first to detect conflicts
        lock table scan stats: seeked 1 times (1 internal); stepped 0 times (0 internal)
        lock table scan complete without conflicts; dropping latches early
        executing read-only batch
        evaluated Scan command header:<key:"\376\222\362\211" end_key:"\376\222\362\212" > scan_format:BATCH_RESPONSE , txn="sql txn" meta={id=5406ea35 key=/Min iso=Serializable pri=0.02132594 epo=0 ts=1764090355.270054262,0 min=1764090355.270054262,0 seq=0} lock=false stat=PENDING rts=1764090355.270054262,0 gul=1764090355.770054262,0 obs={n1@1764090355.272016864,0} : resp=header:<> col_batches:<> , err=<nil>
        n1 scan stats: stepped 0 times (0 internal); seeked 1 times (1 internal); block-bytes: (total 0 B, cached 0 B, duration 0s); points: (count 0, key-bytes 0 B, value-bytes 0 B, tombstoned: 0) ranges: (count 0), (contained-points 0, skipped-points 0) evaluated requests: 0 gets, 1 scans, 0 reverse scans
        read completed
        node sending response
        received batch response
        proxy result is (err: <nil>), *kvpb.ScanResponse <nil>
        proxy request completed
        node sending response
        received batch response
        r_u:0.625 
        recording span to refresh: /Tenant/10/Table/106/{1-2}
        ComponentStats{ID: {40c121db-1f6b-4486-9369-a031c38f81e6 PROCESSOR 0 2 }, KV nodes: n1, KV regions: test, KV time: 5ms, KV rows decoded: 0, KV bytes read: 0 B, KV gRPC calls: 1, KV pairs read: 0, MVCC step count (ext/int): 0/0, MVCC seek count (ext/int): 1/1, max memory allocated: 20 KiB, sql cpu time: 17µs, batches output: 0, rows output: 0}
        ComponentStats{ID: {00000000-0000-0000-0000-000000000000 UNSET 0 ? }}
        ComponentStats{ID: {40c121db-1f6b-4486-9369-a031c38f81e6 FLOW 0 2 test}}
        execution ends
        rows affected: 0
        === SPAN START: commit sql txn ===
        AutoCommit. err: <nil>
        releasing 1 descriptors
        release: 106("foo",010180ec7473b5a83543679f3d5fce668684cd) ver=1:0,0, refcount=0
        [NoTxn pos:12] executing ExecStmt: SET TRACING = off
        executing: SET TRACING = off in state: NoTxn
--- FAIL: TestTrace/ShowTraceForVectorized/TracingOn/node-1 (0.03s)
=== RUN   TestTrace/Session
--- FAIL: TestTrace/Session (0.55s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTrace.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

sql.TestTrace [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20788916?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20788916?buildTab=artifacts#/) on master @ [309577c52479ee3622bcb63ffe0f051a3bc349e2](https://github.com/cockroachdb/cockroach/commits/309577c52479ee3622bcb63ffe0f051a3bc349e2):


```
        creating DistSQL plan with isLocal=true
        noMutations = true
        containsLocking = false
        running DistSQL plan
        === SPAN START: flow ===
        starting (0 processors, 0 startables) asynchronously
        running *rowexec.tableReader in the flow's goroutine
        === SPAN START: table reader ===
        starting scan with parallelize=false
        Scan /Table/106/{1-2}
        kv fetcher: sending a batch with 1 requests
        === SPAN START: txn coordinator send ===
        === SPAN START: dist sender send ===
        querying next range at /Table/106/1
        key: /Table/106/1, desc: r79:/Table/{76-107/1/5} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=5]
        r79: sending batch 1 Scan to (n1,s1):1
        sending request to 127.0.0.1:39493
        sending batch request
        === SPAN START: /cockroach.roachpb.KVBatch/Batch ===
        === SPAN START: /cockroach.roachpb.Internal/Batch ===
        node received request: 1 Scan
        executing Scan [/Table/106/1,/Table/106/2), [txn: f30fbc59], [can-forward-ts], [max_span_request_keys: 0], [target_bytes: 14331]
        read-only path
        sequencing request
        acquiring latches
        scanning lock table for conflicting locks
        can drop latches early for batch (Scan [/Table/106/1,/Table/106/2), [txn: f30fbc59], [can-forward-ts], [max_span_request_keys: 0], [target_bytes: 14331]); scanning lock table first to detect conflicts
        lock table scan stats: seeked 1 times (1 internal); stepped 0 times (0 internal)
        lock table scan complete without conflicts; dropping latches early
        executing read-only batch
        evaluated Scan command header:<key:"\362\211" end_key:"\362\212" > scan_format:BATCH_RESPONSE , txn="sql txn" meta={id=f30fbc59 key=/Min iso=Serializable pri=0.02612048 epo=0 ts=1764323211.456026800,0 min=1764323211.456026800,0 seq=0} lock=false stat=PENDING rts=1764323211.456026800,0 gul=1764323211.956026800,0 obs={n1@1764323211.457144345,0 n3@1764323211.456026800,0} : resp=header:<> col_batches:<> , err=<nil>
        n1 scan stats: stepped 0 times (0 internal); seeked 1 times (1 internal); block-bytes: (total 0 B, cached 0 B, duration 0s); points: (count 0, key-bytes 0 B, value-bytes 0 B, tombstoned: 0) ranges: (count 0), (contained-points 0, skipped-points 0) evaluated requests: 0 gets, 1 scans, 0 reverse scans
        read completed
        node sending response
        received batch response
        recording span to refresh: /Table/106/{1-2}
        ComponentStats{ID: {efe0b720-122c-4b1f-b0f3-7b0b569afa0c PROCESSOR 0 3 }, KV nodes: n1, KV regions: test, KV time: 1ms, KV rows decoded: 0, KV bytes read: 0 B, KV gRPC calls: 1, KV pairs read: 0, MVCC step count (ext/int): 0/0, MVCC seek count (ext/int): 1/1, rows output: 0}
        ComponentStats{ID: {efe0b720-122c-4b1f-b0f3-7b0b569afa0c FLOW 0 3 test}}
        execution ends
        rows affected: 0
        === SPAN START: commit sql txn ===
        AutoCommit. err: <nil>
        releasing 4 descriptors
        release: 100("defaultdb",010180fcc29583353a4148918ff581c55ddfdb) ver=1:0,0, refcount=0
        release: 104("test",010180fcc29583353a4148918ff581c55ddfdb) ver=1:0,0, refcount=0
        release: 105("public",010180fcc29583353a4148918ff581c55ddfdb) ver=1:0,0, refcount=0
        release: 106("foo",010180fcc29583353a4148918ff581c55ddfdb) ver=1:0,0, refcount=0
        [NoTxn pos:10] executing ExecStmt: SET TRACING = off
        executing: SET TRACING = off in state: NoTxn
--- FAIL: TestTrace/Session/TracingOff/node-2 (0.04s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTrace.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

sql.TestTrace [failed](https://mesolite.cluster.engflow.com/invocations/default/67c3b4a6-febe-41c8-afd4-55e89f7bb1de?testReportRun=1&testReportShard=16&testReportAttempt=1#targets-Ly9wa2cvc3FsOnNxbF90ZXN0) on master @ [0f2bc7b8fb79175f32fe44754a872845cd01e038](https://github.com/cockroachdb/cockroach/commits/0f2bc7b8fb79175f32fe44754a872845cd01e038):


```
        acquiring latches
        scanning lock table for conflicting locks
        can't serve follower read; closed timestamp too low by: 3.580037833s; maxClosed: 1764397442.366956255,0 ts: 1764397445.446994088,0 uncertaintyLimit: 1764397445.946994088,0
        replica.Send got error: [NotLeaseHolderError] lease held by different store; r79: replica (n3,s3):2 not lease holder; current lease is repl=(n1,s1):1 seq=1 start=0,0 type=LeaseEpoch epo=1 min-exp=0,0 pro=1764397440.547047718,0 acq=Request
        error from stores.Send: [NotLeaseHolderError] lease held by different store; r79: replica (n3,s3):2 not lease holder; current lease is repl=(n1,s1):1 seq=1 start=0,0 type=LeaseEpoch epo=1 min-exp=0,0 pro=1764397440.547047718,0 acq=Request: "sql txn" meta={id=a09d7419 key=/Min iso=Serializable pri=0.01048236 epo=0 ts=1764397445.446994088,0 min=1764397445.446994088,0 seq=0} lock=false stat=PENDING rts=1764397445.446994088,0 gul=1764397445.946994088,0 obs={n3@1764397445.446994088,0}
        node sending response
        received batch response
        [NotLeaseHolderError] lease held by different store; r79: replica (n3,s3):2 not lease holder; current lease is repl=(n1,s1):1 seq=1 start=0,0 type=LeaseEpoch epo=1 min-exp=0,0 pro=1764397440.547047718,0 acq=Request: "sql txn" meta={id=a09d7419 key=/Min iso=Serializable pri=0.01048236 epo=0 ts=1764397445.446994088,0 min=1764397445.446994088,0 seq=0} lock=false stat=PENDING rts=1764397445.446994088,0 gul=1764397445.946994088,0 obs={n3@1764397445.446994088,0}
        backing off due to NotLeaseHolderErr with stale info (updatedLH=false intentionallySentToFollower=false leaseholderUnavailable=false)
        will retry after 52.687478ms
        application error: [NotLeaseHolderError] lease held by different store; r79: replica (n3,s3):2 not lease holder; current lease is repl=(n1,s1):1 seq=1 start=0,0 type=LeaseEpoch epo=1 min-exp=0,0 pro=1764397440.547047718,0 acq=Request: "sql txn" meta={id=a09d7419 key=/Min iso=Serializable pri=0.01048236 epo=0 ts=1764397445.446994088,0 min=1764397445.446994088,0 seq=0} lock=false stat=PENDING rts=1764397445.446994088,0 gul=1764397445.946994088,0 obs={n3@1764397445.446994088,0}
        trying next peer (n1,s1):1
        sending request to 127.0.0.1:44207
        sending batch request
        === SPAN START: /cockroach.roachpb.KVBatch/Batch ===
        === SPAN START: /cockroach.roachpb.Internal/Batch ===
        node received request: 1 Scan
        executing Scan [/Table/106/1,/Table/106/2), [txn: a09d7419], [can-forward-ts], [max_span_request_keys: 0], [target_bytes: 49589]
        read-only path
        sequencing request
        acquiring latches
        scanning lock table for conflicting locks
        can drop latches early for batch (Scan [/Table/106/1,/Table/106/2), [txn: a09d7419], [can-forward-ts], [max_span_request_keys: 0], [target_bytes: 49589]); scanning lock table first to detect conflicts
        lock table scan stats: seeked 1 times (1 internal); stepped 0 times (0 internal)
        lock table scan complete without conflicts; dropping latches early
        executing read-only batch
        evaluated Scan command header:<key:"\362\211" end_key:"\362\212" > scan_format:BATCH_RESPONSE , txn="sql txn" meta={id=a09d7419 key=/Min iso=Serializable pri=0.01048236 epo=0 ts=1764397445.446994088,0 min=1764397445.446994088,0 seq=0} lock=false stat=PENDING rts=1764397445.446994088,0 gul=1764397445.946994088,0 obs={n1@1764397445.501816474,0 n3@1764397445.446994088,0} : resp=header:<> col_batches:<> , err=<nil>
        n1 scan stats: stepped 0 times (0 internal); seeked 1 times (1 internal); block-bytes: (total 0 B, cached 0 B, duration 0s); points: (count 0, key-bytes 0 B, value-bytes 0 B, tombstoned: 0) ranges: (count 0), (contained-points 0, skipped-points 0) evaluated requests: 0 gets, 1 scans, 0 reverse scans
        read completed
        node sending response
        received batch response
        recording span to refresh: /Table/106/{1-2}
        ComponentStats{ID: {d071412d-4b3b-44e2-833c-d1fc1febd5c2 PROCESSOR 0 3 }, KV nodes: n1, KV regions: test, KV time: 55ms, KV rows decoded: 0, KV bytes read: 0 B, KV gRPC calls: 1, KV pairs read: 0, MVCC step count (ext/int): 0/0, MVCC seek count (ext/int): 1/1, rows output: 0}
        ComponentStats{ID: {d071412d-4b3b-44e2-833c-d1fc1febd5c2 FLOW 0 3 test}}
        execution ends
        rows affected: 0
        === SPAN START: commit sql txn ===
        AutoCommit. err: <nil>
        releasing 4 descriptors
        release: 100("defaultdb",0101805cd7f744e92a466cb25677fbd3ab3cd0) ver=1:0,0, refcount=0
        release: 104("test",0101805cd7f744e92a466cb25677fbd3ab3cd0) ver=1:0,0, refcount=0
        release: 105("public",0101805cd7f744e92a466cb25677fbd3ab3cd0) ver=1:0,0, refcount=0
        release: 106("foo",0101805cd7f744e92a466cb25677fbd3ab3cd0) ver=1:0,0, refcount=0
        [NoTxn pos:10] executing ExecStmt: SET TRACING = off
        executing: SET TRACING = off in state: NoTxn
--- FAIL: TestTrace/Session/TracingOn/node-2 (0.07s)
=== RUN   TestTrace/ShowTraceFor
--- FAIL: TestTrace/ShowTraceFor (0.29s)
=== RUN   TestTrace/ShowTraceFor/TracingOn
--- FAIL: TestTrace/ShowTraceFor/TracingOn (0.14s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=16</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTrace.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

sql.TestTrace [failed](https://mesolite.cluster.engflow.com/invocations/default/706a1673-2c9f-43d5-a360-df046da32181?testReportRun=1&testReportShard=16&testReportAttempt=1#targets-Ly9wa2cvc3FsOnNxbF90ZXN0) on master @ [0f2bc7b8fb79175f32fe44754a872845cd01e038](https://github.com/cockroachdb/cockroach/commits/0f2bc7b8fb79175f32fe44754a872845cd01e038):


```
        scanning lock table for conflicting locks
        can't serve follower read; closed timestamp too low by: 3.657063903s; maxClosed: 1764482629.362943557,0 ts: 1764482632.520007460,0 uncertaintyLimit: 1764482633.020007460,0
        replica.Send got error: [NotLeaseHolderError] lease held by different store; r92: replica (n2,s2):2 not lease holder; current lease is repl=(n1,s1):1 seq=1 start=0,0 type=LeaseEpoch epo=1 min-exp=0,0 pro=1764482626.308732071,0 acq=Request
        proxy request for Scan [/Tenant/10/Table/106/1,/Tenant/10/Table/106/2), [txn: eed59d08], [can-forward-ts], [max_span_request_keys: 0], [target_bytes: 10485760] after local error [NotLeaseHolderError] lease held by different store; r92: replica (n2,s2):2 not lease holder; current lease is repl=(n1,s1):1 seq=1 start=0,0 type=LeaseEpoch epo=1 min-exp=0,0 pro=1764482626.308732071,0 acq=Request: "sql txn" meta={id=eed59d08 key=/Min iso=Serializable pri=0.04310790 epo=0 ts=1764482632.520007460,0 min=1764482632.520007460,0 seq=0} lock=false stat=PENDING rts=1764482632.520007460,0 gul=1764482633.020007460,0 obs={n2@1764482632.521449281,0}
        === SPAN START: dist sender send ===
        processing a proxy request to repl=(n1,s1):1 seq=1 start=0,0 type=LeaseEpoch epo=1 min-exp=0,0 pro=1764482626.308732071,0 acq=Request
        clearing entries overlapping r92:/Tenant/10/Table/{48-107/1/5} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=17]
        sending request to 127.0.0.1:45935
        sending batch request
        === SPAN START: /cockroach.roachpb.KVBatch/Batch ===
        === SPAN START: /cockroach.roachpb.Internal/Batch ===
        node received request: 1 Scan
        executing Scan [/Tenant/10/Table/106/1,/Tenant/10/Table/106/2), [txn: eed59d08], [can-forward-ts], [max_span_request_keys: 0], [target_bytes: 10485760]
        read-only path
        sequencing request
        acquiring latches
        scanning lock table for conflicting locks
        can drop latches early for batch (Scan [/Tenant/10/Table/106/1,/Tenant/10/Table/106/2), [txn: eed59d08], [can-forward-ts], [max_span_request_keys: 0], [target_bytes: 10485760]); scanning lock table first to detect conflicts
        lock table scan stats: seeked 1 times (1 internal); stepped 0 times (0 internal)
        lock table scan complete without conflicts; dropping latches early
        executing read-only batch
        evaluated Scan command header:<key:"\376\222\362\211" end_key:"\376\222\362\212" > scan_format:BATCH_RESPONSE , txn="sql txn" meta={id=eed59d08 key=/Min iso=Serializable pri=0.04310790 epo=0 ts=1764482632.520007460,0 min=1764482632.520007460,0 seq=0} lock=false stat=PENDING rts=1764482632.520007460,0 gul=1764482633.020007460,0 obs={n1@1764482632.522173461,0} : resp=header:<> col_batches:<> , err=<nil>
        n1 scan stats: stepped 0 times (0 internal); seeked 1 times (1 internal); block-bytes: (total 0 B, cached 0 B, duration 0s); points: (count 0, key-bytes 0 B, value-bytes 0 B, tombstoned: 0) ranges: (count 0), (contained-points 0, skipped-points 0) evaluated requests: 0 gets, 1 scans, 0 reverse scans
        read completed
        node sending response
        received batch response
        proxy result is (err: <nil>), *kvpb.ScanResponse <nil>
        proxy request completed
        node sending response
        received batch response
        r_u:0.625 
        recording span to refresh: /Tenant/10/Table/106/{1-2}
        ComponentStats{ID: {db62318d-a5de-489a-aece-433dc17116d3 PROCESSOR 0 2 }, KV nodes: n1, KV regions: test, KV time: 3ms, KV rows decoded: 0, KV bytes read: 0 B, KV gRPC calls: 1, KV pairs read: 0, MVCC step count (ext/int): 0/0, MVCC seek count (ext/int): 1/1, rows output: 0}
        ComponentStats{ID: {db62318d-a5de-489a-aece-433dc17116d3 FLOW 0 2 test}}
        execution ends
        rows affected: 0
        === SPAN START: commit sql txn ===
        AutoCommit. err: <nil>
        releasing 4 descriptors
        release: 100("defaultdb",010180ad8e1c7341cd4f54ab1fed3a447754b4) ver=1:0,0, refcount=0
        release: 104("test",010180ad8e1c7341cd4f54ab1fed3a447754b4) ver=1:0,0, refcount=0
        release: 105("public",010180ad8e1c7341cd4f54ab1fed3a447754b4) ver=1:0,0, refcount=0
        release: 106("foo",010180ad8e1c7341cd4f54ab1fed3a447754b4) ver=1:0,0, refcount=0
        [NoTxn pos:10] executing ExecStmt: SET TRACING = off
        executing: SET TRACING = off in state: NoTxn
--- FAIL: TestTrace/ShowTraceFor/TracingOn/node-1 (0.03s)
=== RUN   TestTrace/ShowTraceFor
--- FAIL: TestTrace/ShowTraceFor (0.12s)
=== RUN   TestTrace/ShowTraceForVectorized
--- FAIL: TestTrace/ShowTraceForVectorized (0.12s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=16</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTrace.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/server/BUILD.bazel
+++ b/pkg/server/BUILD.bazel
@@ -358,6 +358,7 @@ go_library(
         "//pkg/util/timeutil/ptp",
         "//pkg/util/tracing",
         "//pkg/util/tracing/collector",
+        "//pkg/util/tracing/drpcinterceptor",
         "//pkg/util/tracing/goexectrace",
         "//pkg/util/tracing/service",
         "//pkg/util/tracing/tracingpb",

--- a/pkg/server/node.go
+++ b/pkg/server/node.go
@@ -74,6 +74,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/cockroach/pkg/util/tracing"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing/drpcinterceptor"
 	"github.com/cockroachdb/cockroach/pkg/util/tracing/tracingpb"
 	"github.com/cockroachdb/cockroach/pkg/util/tracing/tracingutil"
 	"github.com/cockroachdb/cockroach/pkg/util/unique"
@@ -2018,12 +2019,20 @@ func setupSpanForIncomingRPC(
 ) (context.Context, spanForRequest) {
 	var newSpan *tracing.Span
 	remoteParent := !ba.TraceInfo.Empty()
+	// DRPC server interceptors inject the method name into the context
+	// because DRPC uses different method names (/cockroach.roachpb.KVBatch/Batch).
+	// For gRPC and local requests, we fall back to tracingutil.BatchMethodName
+	// (/cockroach.roachpb.Internal/Batch), so gRPC interceptors don't need to inject it.
+	methodName := tracingutil.BatchMethodName
+	if drpcinterceptor.IsDRPCRequest(ctx) {
+		methodName = tracingutil.KVBatchMethodName
+	}
 	if !remoteParent {
 		// This is either a local request which circumvented gRPC, or a remote
 		// request that didn't specify tracing information. We make a child span
 		// if the incoming request would like to be traced.
 		ctx, newSpan = tracing.ChildSpan(ctx,
-			tracingutil.BatchMethodName, tracing.WithServerSpanKind)
+			methodName, tracing.WithServerSpanKind)
 	} else {
 		// Non-local call. Tracing information comes from the request proto.
 
@@ -2035,7 +2044,7 @@ func setupSpanForIncomingRPC(
 		}
 
 		ctx, newSpan = tr.StartSpanCtx(
-			ctx, tracingutil.BatchMethodName,
+			ctx, methodName,
 			tracing.WithRemoteParentFromTraceInfo(ba.TraceInfo),
 			tracing.WithServerSpanKind)
 	}

--- a/pkg/sql/trace_test.go
+++ b/pkg/sql/trace_test.go
@@ -52,7 +52,7 @@ func TestTrace(t *testing.T) {
 		"local proposal",
 		"admissionWorkQueueWait",
 		"index recommendation",
-		"/cockroach.roachpb.KVBatch/Batch", // present with dRPC, absent with gRPC
+		"/cockroach.roachpb.Internal/Batch",
 	}
 	// Depending on whether the data is local or not, we may not see these
 	// spans. Only applicable with distsql=on.
@@ -67,6 +67,12 @@ func TestTrace(t *testing.T) {
 		"/cockroach.roachpb.Internal/RangeLookup",
 		"executeWriteBatch",
 	}
+	// KVBatch spans are optional since they may or may not appear depending
+	// on whether DRPC is enabled and which nodes handle the request.
+	drpcOptionalSpans := []string{
+		"/cockroach.roachpb.KVBatch/Batch",
+		"/cockroach.roachpb.KVBatch/BatchStream",
+	}
 	commonExpSpans := []string{
 		"session recording",
 		"sql txn",
@@ -76,7 +82,6 @@ func TestTrace(t *testing.T) {
 		"consuming rows",
 		"txn coordinator send",
 		"dist sender send",
-		"/cockroach.roachpb.Internal/Batch",
 		"commit sql txn",
 	}
 	nonVectorizedExpSpans := append(append([]string(nil), commonExpSpans...), []string{
@@ -206,6 +211,11 @@ func TestTrace(t *testing.T) {
 
 	for _, test := range testData {
 		optionalSpans := append([]string{}, alwaysOptionalSpans...)
+		// Check if DRPC is enabled. When DRPC is enabled, remote KV requests
+		// produce DRPC-style span names (/cockroach.roachpb.KVBatch/Batch).
+		if cluster.IsDRPCEnabled() {
+			optionalSpans = append(optionalSpans, drpcOptionalSpans...)
+		}
 		if test.distSQL == "on" {
 			optionalSpans = append(optionalSpans, distsqlOptionalSpans...)
 		}

--- a/pkg/testutils/serverutils/test_cluster_shim.go
+++ b/pkg/testutils/serverutils/test_cluster_shim.go
@@ -248,6 +248,10 @@ type TestClusterInterface interface {
 	// default tenant for testing.
 	StartedDefaultTestTenant() bool
 
+	// IsDRPCEnabled returns whether DRPC is enabled for inter-node
+	// communication in this cluster.
+	IsDRPCEnabled() bool
+
 	// ApplicationLayer calls .ApplicationLayer() on the ith server in
 	// the cluster.
 	ApplicationLayer(idx int) ApplicationLayerInterface

--- a/pkg/testutils/serverutils/test_server_shim.go
+++ b/pkg/testutils/serverutils/test_server_shim.go
@@ -680,7 +680,6 @@ func ShouldEnableDRPC(
 		}
 		return base.TestDRPCEnabled
 	}
-
 	return base.TestDRPCDisabled
 }
 

--- a/pkg/testutils/testcluster/testcluster.go
+++ b/pkg/testutils/testcluster/testcluster.go
@@ -130,6 +130,12 @@ func (tc *TestCluster) StartedDefaultTestTenant() bool {
 	return tc.Servers[0].TenantController().StartedDefaultTestTenant()
 }
 
+// IsDRPCEnabled returns whether DRPC is enabled for inter-node communication
+// in this cluster.
+func (tc *TestCluster) IsDRPCEnabled() bool {
+	return tc.defaultDRPCOption == base.TestDRPCEnabled
+}
+
 // ApplicationLayer calls .ApplicationLayer() on the ith server in
 // the cluster.
 func (tc *TestCluster) ApplicationLayer(idx int) serverutils.ApplicationLayerInterface {

--- a/pkg/util/tracing/drpcinterceptor/drpc_interceptor.go
+++ b/pkg/util/tracing/drpcinterceptor/drpc_interceptor.go
@@ -22,6 +22,14 @@ import (
 	"storj.io/drpc/drpcmux"
 )
 
+// drpcRequestKey is a context key type used to mark a request as coming via DRPC.
+type drpcRequestKey struct{}
+
+// IsDRPCRequest returns true if the context indicates this request came via DRPC.
+func IsDRPCRequest(ctx context.Context) bool {
+	return ctx.Value(drpcRequestKey{}) != nil
+}
+
 // ExtractSpanMetaFromDRPCCtx retrieves trace context (as a SpanMeta) that has
 // been propagated through DRPC metadata in the context.
 func ExtractSpanMetaFromDRPCCtx(
@@ -45,6 +53,7 @@ func ServerInterceptor(tracer *tracing.Tracer) drpcmux.UnaryServerInterceptor {
 		rpc string,
 		handler drpcmux.UnaryHandler,
 	) (interface{}, error) {
+		ctx = context.WithValue(ctx, drpcRequestKey{}, struct{}{})
 		if tracingutil.MethodExcludedFromTracing(rpc) {
 			return handler(ctx, req)
 		}
@@ -83,6 +92,10 @@ func StreamServerInterceptor(tracer *tracing.Tracer) drpcmux.StreamServerInterce
 		rpc string,
 		handler drpcmux.StreamHandler,
 	) (interface{}, error) {
+		stream = &tracingServerStream{
+			Stream: stream,
+			ctx:    context.WithValue(stream.Context(), drpcRequestKey{}, struct{}{}),
+		}
 		if tracingutil.MethodExcludedFromTracing(rpc) {
 			return handler(stream)
 		}

